### PR TITLE
201381 - setting credentials late [1/3]

### PR DESCRIPTION
**Pull request by vvoicehovics**
### Description of Changes

In this part 1 handle additions to the API. New constructor for config builder `Builder(region, featureSet)`. Feature set tells which bits of the sdk do we want to partially initialise (otherwise we could init one that never gets credentials resulting in caches endlessly growing).

1. credentials given immediately (old constructor). Trying to set credentials later results in an exception. Feature set depends on provided creds
2. new constructor.
- partial init for Optimobile means that events etc are cached. Caches are flushed when credentials given.
- partial init for Optimove means (currently) init when credentials are given. Only then optimove starts doing anything.
- not partially initing here means optimobile/optimove wont do anything, setting creds results in an exception

Additionally tweak test app to allow late setting of creds. Inputs appear only when delayed init is used.

TO COME:
1. flushing caches when credentials arrived
4. refactor early aborts if sdk is not fully initialised
5. cache optimove actions (?)
6. deep links

### Breaking Changes

- None

### Release Checklist

Prepare:

- [ ] Detail any breaking changes. Breaking changes require a new major version number, and a migration guide in wiki / README.md

Bump versions in:

- [ ] CHANGELOG.md
- [ ] gradle.properties
- [ ] add links to newly created wiki pages to readme
- [ ] Update major version numbers in wiki (basic integration + push guides)

### Integration tests

_T&T Only_

- [ ] Init SDK with only optimove credentials
- [ ] Associate customer
- [ ] Associate email
- [ ] Track events

_Mobile Only_

- [ ] Init SDK with all credentials
- [ ] Track events
- [ ] Associate customer (verify both backends)
- [ ] Register for push
- [ ] Opt-in for In-App
- [ ] Send test push
- [ ] Send test In-App
- [ ] Receive / trigger deep link handler (In-App/Push)
- [ ] Receive / trigger the content extension, render image and action buttons for push
- [ ] Verify push opened handler

_Deferred Deep Links_

- [ ] With app installed, trigger deep link handler
- [ ] With app uninstalled, follow deep link, install test bundle, verify deep link read from Clipboard, trigger deep link handler

_Combined_

- [ ] Track event for T&T, verify push received
- [ ] Trigger scheduled campaign, verify push received
- [ ] Trigger scheduled campaign, verify In-App received

### Release Procedure

- [ ] Squash and merge `dev` to `master`
- [ ] Delete branch once merged

### CODE REVIEW

File: OptimoveSDK/app/build.gradle [hunk 1]
```diff
@@ -4,7 +4,7 @@ android {
     compileSdkVersion 31
     buildToolsVersion '30.0.3'
     defaultConfig {
-        applicationId "com.optimove.sdk.optimovemobilesdk"
+        applicationId "com.optimove.android.optimovemobilesdk"
```
Comment (vvoicehovics) [inline OptimoveSDK/app/build.gradle:7]: random error spottd in the test app

File: OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/OptimoveConfig.java [hunk 2]
```diff
@@ -58,6 +58,8 @@ public final class OptimoveConfig {
 
     private @Nullable LogLevel minLogLevel;
 
+    private @Nullable PartialInitType partialInitType;
```
Comment (eligutovsky) [inline OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/OptimoveConfig.java:61]: Recommend to rename it `initialization strategy`

**Comment by k-antipochkin** [comment]
@vvoicehovics Will the late initialization cause `FirebaseMessagingService` issues? the token and the messages could be received at any point. Currently, the SDK is initialized in the Application extending class, which loads before the Service. If the initialization is delayed, will it break the messaging service flows?

File: OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/OptimoveConfig.java [hunk 3]
```diff
@@ -193,31 +247,47 @@ public DeferredDeepLinkHandlerInterface getDeferredDeepLinkHandler() {
     }
 
     public boolean isOptimoveConfigured(){
-        return this.optimoveToken != null && this.configFileName != null ;
+        // TODO: optimove part is not ready for this yet
+        // this.featureSet == FeatureSet.OPTIMOVE_ONLY || this.featureSet == FeatureSet.ALL
+        return this.optimoveToken != null && this.configFileName != null;
```
Comment (k-antipochkin) [inline OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/OptimoveConfig.java:252]: Do we have any benefit from the feature set now? Seems like it is inferred from whether credentials exist or not. Is something else planned for this enum?

Comment (vvoicehovics) [reply OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/OptimoveConfig.java:252]: there are 2 constructors:
1. old one: init(optimove, optimobile) --> featureSet inferred
2. new one: init(region, featureSet) --> users tell us which features do they want, so, we can tell which parts to partially initialize before they give credentials. (Partial init of optimove is not possible in the 1st pass, but optimobile is)

**Pull request merged**